src/appleseed/foundation/math/filtersamplingtable.cpp

```diff
@@ -0,0 +1,93 @@
+
+//
+// This source file is part of appleseed.
+// Visit https://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2019 Esteban Tovagliari, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "filtersamplingtable.h"
+
+// appleseed.foundation headers.
+#include "foundation/math/scalar.h"
+
+// Standard headers.
+#include <algorithm>
+
+using namespace std;
+
+namespace foundation
+{
+
+FilterSamplingTable::FilterSamplingTable(const BoxFilter1<float>& filter)
+{
+    // We only sample the positive side of the filter.
+    // The negative part is generated by reflection.
+    for (size_t i = 0; i < TableSize; ++i)
+    {
+        const float x = i / static_cast<float>(TableSize);
+        m_table[i] = filter.sample(x * 0.5f + 0.5f);
+    }
+
+    m_table[TableSize] = filter.get_radius();
+}
+
+FilterSamplingTable::FilterSamplingTable(const TriangleFilter1<float>& filter)
+{
+    // We only sample the positive side of the filter.
+    // The negative part is generated by reflection.
+    for (size_t i = 0; i < TableSize; ++i)
+    {
+        const float x = i / static_cast<float>(TableSize);
+        m_table[i] = filter.sample(x * 0.5f + 0.5f);
+    }
+
+    m_table[TableSize] = filter.get_radius();
+}
+
+float FilterSamplingTable::operator[](const size_t i) const
+{
+    assert(i < TableSize);
+    return m_table[i];
+}
+
+float FilterSamplingTable::sample(const float s) const
+{
+    if (s < 0.5f)
+        return -do_sample(2.0f * s);
+
+    return do_sample(2.0f * (s - 0.5f));
+}
+
+float FilterSamplingTable::do_sample(const float s) const
+{
+    assert(s >= 0.0f);
+    assert(s < 1.0f);
+
+    size_t i;
+    const float t = floor_frac(s * TableSize, i);
```

Comment:
It cannot happen, We assert than s is in range, and we pad the table with an extra item to avoid the branch to check that after rounding errors i becomes TableSize.

Reply:
Ok.